File: frontend/src/types/graph.ts
/** Graph node from backend API */
export interface GraphNode {
  id: string
  name: string
  category: string | null
  level: number
  level_label: 'none' | 'beginner' | 'intermediate' | 'advanced'
  symbolSize: number
  itemStyle?: { color: string }
  label?: { show: boolean; formatter: string }
}

/** Graph edge from backend API */
export interface GraphEdge {
  source: string
  target: string
  relation_type: 'PREREQUISITE' | 'INCLUDES' | 'SIMILAR' | 'COMPLEMENTARY'
  weight: number
  lineStyle: {
    color: string
    width: number
    type: 'solid' | 'dashed' | 'dotted'
    curveness: number
    opacity?: number
  }
}

/** Category with color (replaces frontend hardcoded categoryColors) */
export interface CategoryItem {
  name: string
  color: string
}

/** Gap skill from backend API */
export interface GapSkill {
  skill_name: string
  requirement_level: 'MUST' | 'NICE' | 'BONUS'
}

/** Graph result (job or user graph) */
export interface GraphResult {
  nodes: GraphNode[]
  edges: GraphEdge[]
  state: 'empty' | 'ready'
  categories: CategoryItem[]
  sunburst_data?: SunburstNode | null
}

/** User graph result with optional gap skills */
export interface UserGraphResult extends GraphResult {
  gap_skills: GapSkill[]
}

/** Occupation role for role selector */
export interface OccupationRole {
  id: number
  name: string
  category: string | null
  description: string | null
}

/** Sunburst hierarchical node for ECharts 旭日图 */
export interface SunburstNode {
  name: string
  value?: number
  id?: string
  itemStyle?: { color: string }
  children?: SunburstNode[]
}

File: backend/app/models/schemas/graph.py
"""Graph module Pydantic request/response models.
"""
from pydantic import BaseModel


class GraphNode(BaseModel):
    id: str
    name: str
    category: str | None = None
    level: float = 0.0
    level_label: str = "none"          # "none" | "beginner" | "intermediate" | "advanced"
    symbolSize: int = 20
    itemStyle: dict | None = None
    importance: float = 3.0            # Job skill importance (1-5)
    required_level: str = "NICE"       # Job skill requirement level (MUST/NICE/BONUS)


class GraphEdge(BaseModel):
    source: str
    target: str
    relation_type: str
    weight: float = 0.5
    lineStyle: dict = {}               # Always populated by backend


class CategoryItem(BaseModel):
    name: str
    color: str


class GraphResult(BaseModel):
    nodes: list[GraphNode] = []
    edges: list[GraphEdge] = []
    state: str = "ready"               # "empty" | "ready"
    categories: list[CategoryItem] = [] # name -> color for frontend
    sunburst_data: dict | None = None   # Hierarchical tree for ECharts sunburst 旭日图 series


class UserGraphResult(GraphResult):
    gap_skills: list = []


class GapSkill(BaseModel):
    skill_name: str
    requirement_level: str
